magic_bridge/ic/src/magic_bridge/src/dab.rs

```diff
@@ -0,0 +1,124 @@
+use std::{str::{FromStr}};
+use ic_kit::{
+  candid::{ CandidType, Deserialize},
+  Principal,
+  ic,
+};
+
+use crate::{types::*, magic::STATE};
+use crate::factory::{CreateCanisterParam};
+
+const DAB_TOKEN_ADDRESS: &str = "xmt67-gqaaa-aaaaa-aahja-cai";
+const MAX_DAB_RETRIES: u8 = 10;
+
+#[derive(CandidType, Deserialize, Clone, PartialEq, Debug)]
+pub enum DetailValue {
+    True,
+    False,
+    U64(u64),
+    I64(i64),
+    Float(f64),
+    Text(String),
+    Principal(Principal),
+    #[serde(with = "serde_bytes")]
+    Slice(Vec<u8>),
+    Vec(Vec<DetailValue>),
+}
+#[derive(CandidType, Deserialize)]
+pub struct DABParams {
+    pub name: String,
+    pub description: String,
+    pub thumbnail: String,
+    pub frontend: Option<String>,
+    pub principal_id: CanisterId,
+    pub details: Vec<(String, DetailValue)>,
+}
+
+#[derive(CandidType, Debug, Deserialize)]
+pub enum OperationError {
+    NotAuthorized,
+    NonExistentItem,
+    BadParameters,
+    Unknown(String),
+}
+
+pub type DABResponse = Result<(), OperationError>;
+
+/*
+    Try to register all canisters given in the failed_canisters parameter.
+    Returns a vector of canisters that failed to register.
+    If the registration fails, it increases the retry_count for the canister.
+*/
+pub async fn retry_failed_canisters(
+    mut failed_canisters: Vec<(Principal, (CreateCanisterParam, u8))>
+    ) -> Vec<(Principal, (CreateCanisterParam, u8))>  {
+        let mut failed_retry_canisters = Vec::new();
+        for (canister_id,(params, retry_count)) in failed_canisters.drain(..) {
```

Comment:
Just curious on this one, if using drain, would be best to just back to the `failed_canisters` allocated space

Reply:
Yeah, that drain wasn't really necessary so I removed it. After all, when this function returns the vector with the remaining failed canisters, the next step is to replace the entire vector for this new one... 
The logic is: 
This retries a given vector of failed canisters, and returns a vector with the ones that remains in fail.
It's up to the caller what to do with the old vector and the one that is returned.
```rust
pub async fn retry_failed_canisters(
    mut failed_canisters: Vec<(Principal, (CreateCanisterParam, RetryCount))>,
) -> Vec<(Principal, (CreateCanisterParam, RetryCount))> {}
```